File: 10. Backtracking/Permutations/sol1.py
class Solution:
    # @param A : list of integers
    # @return a list of list of integers
    def __init__(self):
        self.out=[]
    def permuteHelper(self,soFar,Rest):
        if (len(Rest)==0):
            #print (soFar)
            self.out.append(soFar)
        else:
            for i in range(len(Rest)):
                soFar_new=soFar+[Rest[i]]
                Rest_new=Rest[0:i]+Rest[i+1:]
                self.permuteHelper(soFar_new,Rest_new)
    def permute(self, A):
        self.permuteHelper([],A)
        return self.out
sol_obj=Solution()
A=[1,2,3]
out=sol_obj.permute(A)
print (out)

File: 10. Backtracking/Permutations/sol1.cpp
#include <iostream>
#include <string>
#include <stdio.h>
#include <ctype.h>
#include <vector>
using namespace std;

void permuteHelper(vector<int> &Rest,vector<int> &soFar,vector<vector<int> > &result){
  if (Rest.empty()){
    result.push_back(soFar);
    return;
  }
  else{
    for (int i=0;i<Rest.size();i++){
      vector<int> Rest_new=Rest;
      Rest_new.erase(Rest_new.begin() + i);
      vector<int> soFar_new=soFar;
      soFar_new.push_back(Rest[i]);
      permuteHelper(Rest_new,soFar_new,result);
    }
  }
}


vector<vector<int> > permute(vector<int> &Rest) {
  vector<int> soFar{};
  vector<vector<int> > result;
  permuteHelper(Rest, soFar, result);
  return result;
}


int main(int argc, const char * argv[]) {
  vector<int> A{ 10, 20, 30 };
  vector<vector< int > > out=permute (A);
  // Check outout
  for(int i=0;i<out.size(); i++) {
    for (int j=0;j<out[i].size(); j++)
    cout << out[i][j] << " ";
    cout << endl;
  }
  return 0;
}
